expr/func_removezeroseries_test.go

```diff
@@ -0,0 +1,115 @@
+package expr
+
+import (
+	"math"
+	"testing"
+
+	"github.com/grafana/metrictank/api/models"
+)
+
+func TestRemoveZeroSeriesIfAtLeastOneNonNull(t *testing.T) {
+	testRemoveZeroSeries(
+		0.0, // xFilesFactor
+		[]models.Series{
+			getSeries("a", "some zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+			getSeries("d", "allZeros", allZeros),
+		},
+		[]models.Series{
+			getSeries("a", "some zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+		},
+		t,
+	)
+}
+
+func TestRemoveZeroSeriesAllowNoNulls(t *testing.T) {
+	testRemoveZeroSeries(
+		1.0, // xFilesFactor
+		[]models.Series{
+			getSeries("c", "no zeros", noZeros),
+		},
+		[]models.Series{
+			getSeries("c", "no zeros", noZeros),
+		},
+		t,
+	)
+}
+
+func TestRemoveZeroSeriesAllow30PercentNulls(t *testing.T) {
+	testRemoveZeroSeries(
+		0.3, // xFilesFactor
+		[]models.Series{
+			getSeries("a", "30% zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+			getSeries("c", "no zeros", noZeros),
+			getSeries("d", "all zeros", allZeros),
+		},
+		[]models.Series{
+			getSeries("a", "30% zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+			getSeries("c", "no zeros", noZeros),
+		},
+		t,
+	)
+}
+
+func TestRemoveZeroSeriesAllow70PercentNulls(t *testing.T) {
+	testRemoveZeroSeries(
+		0.7, // xFilesFactor
+		[]models.Series{
+			getSeries("a", "30% zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+			getSeries("c", "no zeros", noZeros),
+			getSeries("d", "all zeros", allZeros),
+		},
+		[]models.Series{
+			getSeries("c", "no zeros", noZeros),
+		},
+		t,
+	)
+}
+
+func TestRemoveZeroSeriesMissingInputXFilesFactor(t *testing.T) {
+	testRemoveZeroSeries(
+		math.NaN(), // xFilesFactor
+		[]models.Series{
+			getSeries("a", "some zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+			getSeries("d", "allZeros", allZeros),
+		},
+		[]models.Series{
+			getSeries("a", "some zeros", a),
+			getSeries("b", "half zeros", halfZeros),
+		},
+		t,
+	)
+}
+
+func testRemoveZeroSeries(xff float64, in []models.Series, out []models.Series, t *testing.T) {
+	f := NewRemoveZeroSeries()
+	f.(*FuncRemoveZeroSeries).in = NewMock(in)
+	f.(*FuncRemoveZeroSeries).xFilesFactor = xff
+
+	inputCopy := models.SeriesCopy(in) // to later verify that it is unchanged
+
+	dataMap := initDataMap(in)
+
+	got, err := f.Exec(dataMap)
+
+	if err := equalOutput(out, got, nil, err); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("DidNotModifyInput", func(t *testing.T) {
+		if err := equalOutput(inputCopy, in, nil, nil); err != nil {
+			t.Fatalf("Input was modified, err = %s", err)
+		}
+	})
+
+	t.Run("DoesNotDoubleReturnPoints", func(t *testing.T) {
+		if err := dataMap.CheckForOverlappingPoints(); err != nil {
+			t.Fatalf("Point slices in datamap overlap, err = %s", err)
+		}
+	})
+}
```

Comment:
You can remove a lot of these newlines. They don't improve clarity and just make the function less likely to fit on a screen.  Try to group lines together that directly deal with eachother, or group "topics" of the function together.  In this suggestion I grouped the setup, working with the data, and validation together.  This stuff is all a matter of taste so there's no hard and fast rule, but having a section of a function with code/blank/code/blank/code/blank should be avoided, same with blank lines at the beginning or ends of functions.

```suggestion
func testRemoveZeroSeries(xff float64, in []models.Series, out []models.Series, t *testing.T) {
	f := NewRemoveZeroSeries()
	f.(*FuncRemoveZeroSeries).in = NewMock(in)
	f.(*FuncRemoveZeroSeries).xFilesFactor = xff

	inputCopy := models.SeriesCopy(in) // to later verify that it is unchanged
	dataMap := initDataMap(in)
	got, err := f.Exec(dataMap)

	if err := equalOutput(out, got, nil, err); err != nil {
		t.Fatal(err)
	}
	t.Run("DidNotModifyInput", func(t *testing.T) {
		if err := equalOutput(inputCopy, in, nil, nil); err != nil {
			t.Fatalf("Input was modified, err = %s", err)
		}
	})
	t.Run("DoesNotDoubleReturnPoints", func(t *testing.T) {
		if err := dataMap.CheckForOverlappingPoints(); err != nil {
			t.Fatalf("Point slices in datamap overlap, err = %s", err)
		}
	})
}
```